### ISSUE TITLE
Common base class or interface for temporal Instant and Interval classes

### ISSUE THREAD

**Issue by navispatial** [navibyte/geospatial#70]
`Instant` definition, partially:

```dart
/// An instant with a time stamp.
@immutable
class Instant with EquatableMixin {
  /// Creates an instant with a [time] stamp.
  const Instant(this.time);

  /// The time stamp of this instant.
  final DateTime time;

  // ...
}
```

`Interval` definition, partially:

```dart
/// An interval with start and end time stamps (one or both can be null = open).
@immutable
class Interval with EquatableMixin {
  /// Creates a closed interval with [start] and [end] time stamps.
  const Interval.closed(this.start, this.end);

  /// Creates an open ended interval with the [start] time stamp.
  const Interval.openEnd(this.start) : end = null;

  /// Creates an open started interval with the [end] time stamp.
  const Interval.openStart(this.end) : start = null;

  /// Creates a fully open interval with [start] and [end] set to null.
  const Interval.open()
      : start = null,
        end = null;

  /// The start time of the interval. If null then the interval is open started.
  final DateTime? start;

  /// The end time of the interval. If null then the interval is open ended.
  final DateTime? end;

  /// A duration (difference) between [start] and [end] if both are available.
  Duration? get duration {
    final s = start;
    final e = end;
    return s != null && e != null ? e.difference(s) : null;
  }

  /// True if the interval is open (both [start] and [end] are null).
  bool get isOpen => start == null && end == null;

  /// True if the interval is closed (both [start] and [end] are non null).
  bool get isClosed => start != null && end != null;

  // ...
}
```

In some places code might accept either `Instant` or `Interval`. Both extend of course `Object`, but maybe more specific super class or interface might be needed.

**Comment by navispatial** [comment]
Other Dart libraries for handling dates, times, instants and intervals, etc.

* [time_machine](https://pub.dev/packages/time_machine), no null-safe support yet?
  * [Instant](https://pub.dev/documentation/time_machine/latest/time_machine/Instant-class.html) 
    * "*Represents an instant on the global timeline, with nanosecond resolution. An Instant has no concept of a particular time zone or calendar: it simply represents a point in time that can be globally agreed-upon.*" 
  * [Interval](https://pub.dev/documentation/time_machine/latest/time_machine/Interval-class.html)
    * `Interval(Instant start, Instant end)`
    * "*The interval includes the start instant and excludes the end instant. The end may equal the start (resulting in an empty interval), but must not be before the start.*" 
  * [DateInterval](https://pub.dev/documentation/time_machine/latest/time_machine/DateInterval-class.html)
    * `DateInterval(LocalDate start, LocalDate end)`  
* [dart_time](https://pub.dev/packages/dart_date)
  * [Interval](https://pub.dev/documentation/dart_date/latest/dart_date/Interval-class.html)
    * `Interval(DateTime start, DateTime end)`
* [date_format](https://pub.dev/packages/date_format)
* [date_field](https://pub.dev/packages/date_field)
* [date_time_picker](https://pub.dev/packages/date_time_picker)
* ....

**Comment by navispatial** [comment]
Planned inheritance hierarchy (all these as abstract classes or interfaces, implementation details to be hidden):
* `Temporal`
  * `Interval` 
    * `Date`
  * `Instant`   

According to previous definitions `Interval` may contains different types of intervals:
* a closed interval with `start` and `end` time stamps
* an open ended interval with the `start` time stamp
* an open started interval with the `end` time stamp
* a fully open interval with `start` and `end` set to null.

The special case `Date` (with year, month, day) would be of the case *a closed interval with `start` and `end` time stamps*. See also a similar class [Date](https://pub.dev/documentation/database/latest/database/Date-class.html) of the [database](https://pub.dev/packages/database) package.

**Comment by navispatial** [comment]
Planned inheritance hierarchy (all these as abstract classes or interfaces, implementation details to be hidden):
* `Temporal`
  * `Interval` 
    * `Date`
  * `Instant`   

According to previous definitions `Interval` may contains different types of intervals:
* a closed interval with `start` and `end` time stamps
* an open ended interval with the `start` time stamp
* an open started interval with the `end` time stamp
* a fully open interval with `start` and `end` set to null.

The special case `Date` (with year, month, day) would be of the case *a closed interval with `start` and `end` time stamps*. See also a similar class [Date](https://pub.dev/documentation/database/latest/database/Date-class.html) of the [database](https://pub.dev/packages/database) package.

**Comment by navispatial** [comment]
Interval maybe should define, that `start` timestamps are included in the interval, but `end` timestamps not.

**Comment by navispatial** [comment]
Joda package: [Interval](http://joda-time.sourceforge.net/apidocs/org/joda/time/Interval.html)

**Comment by navispatial** [comment]
New base class for `Instant` and `Interval`, initial version:

```dart
/// A base class for *temporal* events like *instants* and *intervals*.
abstract class Temporal {
  /// True if this *temporal* object is set to UTC time.
  bool get isUtc;

  /// Returns this *temporal* object in the UTC time zone.
  Temporal toUtc();

  /// A string representation of this *temporal* object.
  ///
  /// If object is an *instant*, returns ISO-8601 full-precision extended format
  /// representation. See `DateTime.toIso8601String()` for reference.
  ///
  /// If object is an interval, returns a string defined as (where "<start>"
  /// and "<end>" refers to ISO-8601 formatted timestamps):
  /// * closed interval: "<start>/<end>"
  /// * open ended interval: "<start>/.."
  /// * open started interval: "../<end>"
  /// * open interval: "../.."
  String toText();

  /// Returns true if this *temporal* event occurs fully after [instant].
  ///
  /// See `DateTime.isAfter` for reference.
  bool isAfterTime(DateTime instant);

  /// Returns true if this occurs fully before [instant].
  ///
  /// See `DateTime.isAfter` for reference.
  bool isBeforeTime(DateTime instant);
}
```

Also `Instant` and `Interval` has been refactored.

**Comment by navispatial** [comment]
Implemented in pre-release `0.8.0-a.7` of the upcoming BETA-version `0.8.0`
